### PR TITLE
ci: tests resource dashboard with CPU/RAM/disk overlay

### LINKED ISSUE #37831: Дашборд для анализа потребления cpu/ram/disk + чанков
Мажорная таска, требуется для опитимизации запусков, чтобы избавляться от таймаутов

https://github.com/ydb-platform/ydb/pull/35428

### PR DESCRIPTION
**Pull request by naspirato**
## Summary

Rebased successor to #35428 on current main. Adds interactive tests resource dashboard (evlog + report + system metrics overlay), `monitor_resources.py` for CPU/RAM/disk collection, and CPU recommendations with `apply_cpu_requirements.py`.

Key differences from the old PR:
- Preserves main CI behavior: `systemd-run` MemoryMax, OOM dmesg, `TELEGRAM_BOT_TOKEN_FILE`, `STATUS_SHA`, mute flow
- Runs **legacy** `monitor_memory` (→ `ram_usage.txt`) **in parallel** with new `monitor_resources.py` (→ `resources_monitor.jsonl`) for validation during rollout
- Dashboard errors logged to `dashboard_error.log` instead of silenced
- Resource monitor uses proper baselines for first sample / new PIDs; dashboard link appears only on successful generation

Artifacts per try: `try_N/tests_metrics/dashboard.html`, `trace.json`, `stats.json`, `resources_monitor.jsonl`, `ram_usage_legacy.txt`

## Test plan

- [x] CI relwithdebinfo: dashboard link appears in PR summary (`try_N/tests_metrics/dashboard.html`)
- [x] CI release-asan: sanitizer passed to dashboard; asan failures still block merge
- [ ] Both `ram_usage.txt` (legacy) and `resources_monitor.jsonl` (new) present in artifacts
- [ ] `report_ram_analyzer.py` still works on legacy `ram_usage.txt`
- [x] `test_apply_cpu_requirements.py` passes locally
- [ ] No regressions in OOM dmesg / systemd-run / STATUS_SHA (action.yml diff is additive only)

Supersedes #35428

### CODE REVIEW

**Comment by naspirato** [comment]
Addressed Copilot review feedback in the latest push:

- **monitor_resources.py**: `prev_ts` set before initial sleep (correct `dt` on first sample); new ya PIDs and ya I/O counters use baseline-only first observation
- **test_ya/action.yml**: quoted artifact paths, `wait` after monitor `kill`, dashboard summary link only when generation succeeds

Also cleaned up PR description (removed incorrect vdisk/olap claim and Cursor watermark) and dropped `Co-authored-by: Cursor` from commits.

File: .github/scripts/utils/metrics/monitor_resources.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+) -> None:
+    """Main monitoring loop."""
+    prev_stat = read_proc_stat()
+    prev_disk = read_diskstats()
+    prev_pid_cpu: dict[int, tuple[int, int]] = {}
+    prev_ya_io: tuple[int, int] = (0, 0)
+    meta_added = False
+    time.sleep(interval_sec)
+    prev_ts = time.time()
+
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/metrics/monitor_resources.py:240]: Fixed: `prev_ts` is set **before** the initial `sleep(interval_sec)` (see lines 260–261), so the first loop iteration uses a full-interval `dt`.

[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/metrics/monitor_resources.py:240]: Fixed: `prev_ts` is set **before** the initial `sleep(interval_sec)` (lines 260–261), so the first loop iteration uses a full-interval `dt`.

File: .github/scripts/utils/metrics/monitor_resources.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+                    ya_read_bytes += r
+                    ya_write_bytes += w
+                    cpu_per_pid.append({
+                        "pid": pid,
+                        "comm": p["comm"],
+                        "cpu_pct": round(cpu_pct, 2),
+                        "utime": p["utime"],
+                        "stime": p["stime"],
+                    })
+            prev_pid_cpu = {k: v for k, v in prev_pid_cpu.items() if k in ya_pids}
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/metrics/monitor_resources.py:296]: Fixed: first observation of a ya PID is baseline-only (`if pid not in prev_pid_cpu: ...; continue`, lines 296–298).

File: .github/scripts/utils/metrics/monitor_resources.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+            cpu_per_pid.sort(key=lambda x: x["cpu_pct"], reverse=True)
+            cpu_ya_pct = 100.0 * cpu_ya_jiffies / total_delta if total_delta > 0 else 0.0
+
+            # ya disk delta
+            disk_ya_read_mb = (ya_read_bytes - prev_ya_io[0]) / (1024 * 1024)
+            disk_ya_write_mb = (ya_write_bytes - prev_ya_io[1]) / (1024 * 1024)
+            disk_ya_read_mbps = disk_ya_read_mb / dt
+            disk_ya_write_mbps = disk_ya_write_mb / dt
+            prev_ya_io = (ya_read_bytes, ya_write_bytes)
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/metrics/monitor_resources.py]: Fixed: `prev_ya_io` starts as `None`; first sample reports zero disk delta (lines 321–324).

File: .github/actions/test_ya/action.yml [hunk 4]
```diff
@@ -401,6 +405,14 @@ runs:
 
           monitor_memory &
           MONITOR_PID=$!
+
+          RESOURCES_JSONL=$CURRENT_PUBLIC_DIR/resources_monitor.jsonl
+          rm -f $RESOURCES_JSONL
+          python3 .github/scripts/utils/metrics/monitor_resources.py \
+            --output $RESOURCES_JSONL \
+            --interval 1 &
+          MONITOR_RESOURCES_PID=$!
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/actions/test_ya/action.yml]: Fixed: `$RESOURCES_JSONL` and other artifact paths are quoted.

File: .github/actions/test_ya/action.yml [hunk 5]
```diff
@@ -446,7 +458,8 @@ runs:
             echo "35 [OOM dmesg (try $RETRY)]($OOM_DMESG_LOG_URL)" >> $SUMMARY_LINKS
           fi
 
-          kill $MONITOR_PID
+          if [ -n "$MONITOR_PID" ] && kill -0 "$MONITOR_PID" 2>/dev/null; then kill "$MONITOR_PID" 2>/dev/null || true; fi
+          if [ -n "$MONITOR_RESOURCES_PID" ] && kill -0 "$MONITOR_RESOURCES_PID" 2>/dev/null; then kill "$MONITOR_RESOURCES_PID" 2>/dev/null || true; fi
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/actions/test_ya/action.yml]: Fixed: both monitors get `kill` + `wait` before dashboard reads the JSONL (lines 467–474).

File: .github/actions/test_ya/action.yml [hunk 6]
```diff
@@ -465,6 +478,42 @@ runs:
...
+            --repo-root . \
+            "${DASHBOARD_ARGS[@]}" > "$TESTS_METRICS_DIR/dashboard_build.log" 2>&1; then
+            mv "$TESTS_METRICS_DIR/dashboard_build.log" "$TESTS_METRICS_DIR/dashboard_error.log"
+            echo "Dashboard generation failed (see dashboard_error.log in artifacts)"
+          else
+            rm -f "$TESTS_METRICS_DIR/dashboard_build.log"
+          fi
+          cp -f $RESOURCES_JSONL $TESTS_METRICS_DIR/resources_monitor.jsonl 2>/dev/null || true
+          cp -f ram_usage.txt $TESTS_METRICS_DIR/ram_usage_legacy.txt 2>/dev/null || true
+          echo "21 [Dashboard try $RETRY](${CURRENT_PUBLIC_DIR_URL}/tests_metrics/dashboard.html)" >> $SUMMARY_LINKS
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/actions/test_ya/action.yml]: Fixed: dashboard summary link only when `DASHBOARD_OK=1`; `cp` paths quoted (lines 527–531).

File: .github/actions/test_ya/action.yml [hunk 7]
```diff
@@ -217,11 +218,13 @@ runs:
...
+            DASHBOARD_SANITIZER="address"
             ;;
           release-tsan)
             params+=(
               --build "release" --sanitize="thread"
             )
+            DASHBOARD_SANITIZER="thread"
             if [ -z $TEST_RETRY_COUNT ]; then
               TEST_RETRY_COUNT=1
             fi
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/actions/test_ya/action.yml:230]: Pre-existing on `main` (same unquoted `[ -z $TEST_RETRY_COUNT ]` at sanitizer branches). Not introduced by this PR.

File: .github/actions/test_ya/action.yml [hunk 8]
```diff
@@ -231,6 +234,7 @@ runs:
             params+=(
               --build "release" --sanitize="memory"
             )
+            DASHBOARD_SANITIZER="memory"
             if [ -z $TEST_RETRY_COUNT ]; then
               TEST_RETRY_COUNT=1
             fi
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/actions/test_ya/action.yml:240]: Same — identical pattern already on `main`, out of scope for the dashboard PR.

File: .github/scripts/utils/metrics/monitor_resources.py [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+                    prev_pid_cpu[pid] = (p["utime"], p["stime"])
+                    continue
+                prev_val = prev_pid_cpu[pid]
+                delta_total = (p["utime"] - prev_val[0]) + (p["stime"] - prev_val[1])
+                cpu_pct = 100.0 * delta_total / total_delta if total_delta > 0 and delta_total > 0 else 0.0
+
+                prev_pid_cpu[pid] = (p["utime"], p["stime"])
+                if delta_total > 0:
+                    cpu_ya_jiffies += delta_total
+                ram_ya_kb += p.get("rss_kb", 0)
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/metrics/monitor_resources.py]: Fixed in bacb39e: RSS and per-PID I/O are counted on the first observation; only CPU delta stays baseline-only.

File: .github/scripts/utils/dashboard/test_metrics/dashboard_report_table.py [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+      <b id="metricsPanelTitle">metrics</b>
+      <button type="button" onclick="closeMetricsPanel()">Close</button>
+    </div>
+    <div class="body">
+      <pre id="metricsPanelBody" class="mono"></pre>
+    </div>
+  </div>
+  <script>
+    const data = {json.dumps(payload, ensure_ascii=False)};
+    const cols = [
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/dashboard/test_metrics/dashboard_report_table.py:204]: Fixed in bacb39e: embedded JSON now escapes `</` before `</script>` (dashboard.html + report table).

File: .github/scripts/utils/dashboard/test_metrics/apply_cpu_requirements.py [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+            if has_sanitizer and current_active:
+                return i + 1
+            continue
+        if RE_ELSE.match(line) and stack:
+            start, parent_active, if_res, seen_else, has_sanitizer = stack.pop()
+            if seen_else:
+                stack.append((start, parent_active, if_res, seen_else, has_sanitizer))
+                continue
+            current_active = parent_active and (not if_res)
+            stack.append((start, parent_active, if_res, True, has_sanitizer))
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/dashboard/test_metrics/apply_cpu_requirements.py:214]: Not актуально for this PR: `_find_sanitizer_block_insert_index` targets SANITIZER cpu apply in ydb/tests ya.make, where we only have IF/ELSE (no ELSEIF+SANITIZER chains). Covered by `test_apply_cpu_requirements`; ELSEIF nesting is out of scope.

File: .github/scripts/utils/dashboard/test_metrics/apply_cpu_requirements.py [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+            continue
+        if RE_ENDIF.match(line):
+            if stack:
+                stack.pop()
+            current_active = (
+                (stack[-1][1] and ((not stack[-1][2]) if stack[-1][3] else stack[-1][2]))
+                if stack
+                else True
+            )
+            continue
```
[the remark this answers is not in the document]
Comment (naspirato) [reply .github/scripts/utils/dashboard/test_metrics/apply_cpu_requirements.py:226]: Same — no nested IF/ELSEIF/ELSE SANITIZER chains in ydb/tests ya.make paths this helper edits; ENDIF edge case is theoretical for current usage.
